# Lab 3a: Data Types
Welcome to the first half of Lab 3.  
For this lab, be sure to download the zipped file from Canvas, and after unzipping that, move all those files into the same directory as this Jupyter notebook file.

In this lab, you'll first see how to represent and manipulate another fundamental type of data: text.  A piece of text is called a *string* in Python. You'll also see how to invoke *methods*.  A method is very similar to a function, but tied to a particular piece of data. Lastly, you'll begin to see how to work with *collections* of data, like the numbers 2 through 5 or the words "welcome", "to", and "lab".

# Your Name: Nila Manickavasagam

# 1. Review: The building blocks of Python code

The two building blocks of Python code are *expressions* and *statements*.  An **expression** is a piece of code that

* is self-contained, meaning it would make sense to write it on a line by itself, and
* usually has a value.


Here are two expressions that both evaluate to 3

    3
    5 - 2
    
One important form of an expression is the call expression, which first names a function and then describes its arguments. The function returns some value, based on its arguments. Some important mathematical functions are

| Function | Description                                                   |
|----------|---------------------------------------------------------------|
| `abs`      | Returns the absolute value of its argument                    |
| `max`      | Returns the maximum of all its arguments                      |
| `min`      | Returns the minimum of all its arguments                      |
| `pow`      | Raises its first argument to the power of its second argument |
| `round`    | Round its argument to the nearest integer                     |

Here are two call expressions that both evaluate to 3

    abs(2 - 5)
    max(round(2.8), min(pow(2, 10), -1 * pow(2, 10)))

All these expressions but the first are **compound expressions**, meaning that they are actually combinations of several smaller expressions.  `2 + 3` combines the expressions `2` and `3` by addition.  In this case, `2` and `3` are called **subexpressions** because they're expressions that are part of a larger expression.

A **statement** is a whole line of code.  Some statements are just expressions.  The expressions listed above are examples.

Other statements *make something happen* rather than *having a value*.  After they are run, something in the world has changed.  For example, an assignment statement assigns a value to a name. 

A good way to think about this is that we're **evaluating the right-hand side** of the equals sign and **assigning it to the left-hand side**. Here are some assignment statements:
    
    height = 1.3
    the_number_five = abs(-5)
    absolute_height_difference = abs(height - 1.688)

A key idea in programming is that large, interesting things can be built by combining many simple, uninteresting things.  The key to understanding a complicated piece of code is breaking it down into its simple components.

For example, a lot is going on in the last statement above, but it's really just a combination of a few things.  This picture describes what's going on.

<img src="statement.jpg">

**Question 1.1.** In the next cell, assign the name `new_year` to the larger number among the following two numbers:

1. the absolute value of $2^{5}-2^{11}-2^0$, and 
2. $5 \times 13 \times 31$.

Try to use just one statement (one line of code).

In [1]:
new_year = max(abs((2**5)-(2**11)-(2**0)), 5*13*31)
new_year

2017

# 2. Text
Text is one of the most common types of values used in programs. A snippet of text is represented by a **string value** in Python. The word "*string*" is a programming term for a sequence of characters. A string might contain a single character, a word, a sentence, or a whole book.

To distinguish text data from actual code, we demarcate strings by putting quotation marks around them. Single quotes (`'`) and double quotes (`"`) are both valid, but the types of opening and closing quotation marks must match. The contents can be any sequence of characters, including numbers and symbols. 

We've seen strings before in `print` statements.  Below, two different strings are passed as arguments to the `print` function.

In [7]:
print("I <3", 'Data Science')

Just like names can be given to numbers, names can be given to string values.  The names and strings aren't required to be similar in any way. Any name can be assigned to any string.

In [8]:
one = 'two'
plus = '*'
print(one, plus, one)

**Question 2.1.** Yuri Gagarin was the first person to travel through outer space.  When he emerged from his capsule upon landing on Earth, he [reportedly](https://en.wikiquote.org/wiki/Yuri_Gagarin) had the following conversation with a woman and girl who saw the landing:

    The woman asked: "Can it be that you have come from outer space?"
    Gagarin replied: "As a matter of fact, I have!"

The cell below contains unfinished code.  Fill in the `...`s so that it prints out this conversation *exactly* as it appears above.

In [2]:
woman_asking = 'The woman asked:'
woman_quote = '"Can it be that you have come from outer space?"'
gagarin_reply = 'Gagarin replied:'
gagarin_quote = '"As a matter of fact, I have!"'

print(woman_asking, woman_quote)
print(gagarin_reply, gagarin_quote)

The woman asked: "Can it be that you have come from outer space?"
Gagarin replied: "As a matter of fact, I have!"


## 2.1. String Methods

Strings can be transformed using **methods**, which are functions that involve an existing string and some other arguments. One example is the `replace` method, which replaces all instances of some part of a string with some alternative. 

A method is invoked on a string by placing a `.` after the string value, then the name of the method, and finally parentheses containing the arguments. Here's a sketch, where the `<` and `>` symbols aren't part of the syntax; they just mark the boundaries of sub-expressions.

    <expression that evaluates to a string>.<method name>(<argument>, <argument>, ...)

Try to predict the output of these examples, then execute them.

In [3]:
# Replace one letter
'Hello'.replace('o', 'a')

'Hella'

In [4]:
# Replace a sequence of letters, which appears twice
'hitchhiker'.replace('hi', 'ma')

'matchmaker'

Once a name is bound to a string value, methods can be invoked on that name as well. The name doesn't change in this case, so a new name is needed to capture the result. 

In [5]:
sharp = 'edged'
hot = sharp.replace('ed', 'ma')
print('sharp:', sharp)
print('hot:', hot)

sharp: edged
hot: magma


You can call functions on the results of other functions.  For example,

    max(abs(-5), abs(3))

has value 5.  Similarly, you can invoke methods on the results of other method (or function) calls.

In [6]:
# Calling replace on the output of another call to
# replace
'train'.replace('t', 'ing').replace('in', 'de')

'degrade'

Here's a picture of how Python evaluates a "chained" method call like that:

<img src="chaining_method_calls.jpg"/>

**Question 2.1.1.** Assign strings to the names `you` and `this` so that the final expression evaluates to a 10-letter English word with three double letters in a row.

*Hint:* After you guess at some values for `you` and `this`, it's helpful to see the value of the variable `the`.  Try printing the value of `the` by adding a line like this:
    
    print(the)

In [8]:
you = 'keep'
this = 'book'
a = 'beeper'
the = a.replace('p', you) 
the.replace('bee', this)

'bookkeeper'

Other string methods do not take any arguments at all, because the original string is all that's needed to compute the result. In this case, parentheses are still needed, but there's nothing in between the parentheses. Here are some methods that work that way:

|Method name|Value|
|-|-|
|`lower`|a lowercased version of the string|
|`upper`|an uppercased version of the string|
|`capitalize`|a version with the first letter capitalized|
|`title`|a version with the first letter of every word capitalized||


In [9]:
'unIverSITy of tennESSee'.title()

'University Of Tennessee'

All these string methods are useful, but most programmers don't memorize their names or how to use them.  In the "real world," people usually just search the internet for documentation and examples. A complete [list of string methods](https://docs.python.org/3/library/stdtypes.html#string-methods) appears in the Python language documentation. [Stack Overflow](http://stackoverflow.com) has a huge database of answered questions that often demonstrate how to use these methods to achieve various ends.

## 2.2. Converting to and from Strings

Strings and numbers are different *types* of values, even when a string contains the digits of a number. For example, evaluating the following cell causes an error because an integer cannot be added to a string.

In [18]:
8 + "8"

However, there are built-in functions to convert numbers to strings and strings to numbers. 

    int:   Converts a string of digits to an integer ("int") value
    float: Converts a string of digits, perhaps with a decimal point, to a decimal ("float") value
    str:   Converts any value to a string

Try to predict what the following cell will evaluate to, then evaluate it.

In [10]:
8 + int("8")

16

Suppose you're writing a program that looks for dates in a text, and you want your program to find the amount of time that elapsed between two years it has identified.  It doesn't make sense to subtract two texts, but you can first convert the text containing the years into numbers.

**Question 2.2.1.** Finish the code below to compute the number of years that elapsed between `one_year` and `another_year`.  Don't just write the numbers `1618` and `1648` (or `30`); use a conversion function to turn the given text data into numbers.

In [13]:
# Some text data:
one_year = "1618"
another_year = "1648"

# Complete the next line.  Note that we can't just write:
#   another_year - one_year
# If you don't see why, try seeing what happens when you
# write that here.
difference = int(str(another_year)) - int(str(one_year))
difference

30

## 2.3. Strings as function arguments

String values, like numbers, can be arguments to functions and can be returned by functions.  The function `len` takes a single string as its argument and returns the number of characters in the string: its **len**-gth.  

Note that it doesn't count *words*. `len("We are the Tennessee Volunteers")` is 31, not 5.

**Question 2.3.1.**  Use `len` to find out the number of characters in the very long string in the next cell.  (It's the first sentence of the English translation of the 1789 French [Declaration of Rights](http://avalon.law.yale.edu/18th_century/rightsof.asp).)  The length of a string is the total number of characters in it, including things like spaces and punctuation.  Assign `sentence_length` to that number.

In [14]:
a_very_long_sentence = "The representatives of the French people, organized as a National Assembly, believing that the ignorance, neglect, or contempt of the rights of man are the sole cause of public calamities and of the corruption of governments, have determined to set forth in a solemn declaration the natural, unalienable, and sacred rights of man, in order that this declaration, being constantly before all the members of the Social body, shall remind them continually of their rights and duties; in order that the acts of the legislative power, as well as those of the executive power, may be compared at any moment with the objects and purposes of all political institutions and may thus be more respected, and, lastly, in order that the grievances of the citizens, based hereafter upon simple and incontestable principles, shall tend to the maintenance of the constitution and redound to the happiness of all."
sentence_length = len(a_very_long_sentence)
sentence_length

896

# 3. Importing code

Most programming involves work that is very similar to work that has been done before.  Since writing code is time-consuming, it's good to rely on others' published code when you can.  Python allows us to **import** other code, creating a **module** that contains all of the names created by that code.

Python includes many useful modules that are just an `import` away.  We'll look at the `math` module as a first example. The `math` module is extremely useful in computing mathematical expressions in Python. 

Suppose we want to very accurately compute the area of a circle with radius 5 meters.  For that, we need the constant $\pi$, which is roughly 3.14.  Conveniently, the `math` module has `pi` defined for us:

In [24]:
import math
radius = 5
area_of_circle = radius**2 * math.pi
area_of_circle

`pi` is defined inside `math`, and the way that we access names that are inside modules is by writing the module's name, then a dot, then the name of the thing we want:

    <module name>.<name>
    
In order to use a module at all, we must first write the statement `import <module name>`.  That statement creates a module object with things like `pi` in it and then assigns the name `math` to that module.  Above we have done that for `math`.

**Question 3.1.** `math` also provides the name `e` for the base of the natural logarithm, which is roughly 2.71.  Compute $e^{\pi}-\pi$, giving it the name `near_twenty`.

In [15]:
import math
near_twenty = math.e**math.pi - math.pi
near_twenty

19.99909997918947

## 3.1. Importing functions

**Modules** can provide other named things, including **functions**.  For example, `math` provides the name `sin` for the sine function.  Having imported `math` already, we can write `math.sin(3)` to compute the sine of 3.  (Note that the input to this sine function is radians, not degrees; 180 degrees is $\pi$ radians.)

**Question 3.1.1.** A $\frac{\pi}{4}$-radian (45-degree) angle forms a right triangle with equal base and height, pictured below.  If the hypotenuse (the radius of the circle in the picture) is 1, then the height is $\sin(\frac{\pi}{4})$.  Compute that using `sin` and `pi` from the `math` module.  Give the result the name `sine_of_pi_over_four`.

 <img src="http://mathworld.wolfram.com/images/eps-gif/TrigonometryAnglesPi4_1000.gif">

In [16]:
import math
sine_of_pi_over_four = math.sin(math.pi/4)
sine_of_pi_over_four

0.7071067811865476

For your reference, here are some more examples of functions from the `math` module.

Note how different methods take in different number of arguments. Often, the documentation of the module will provide information on how many arguments is required for each method.

In [17]:
# Calculating factorials.
math.factorial(5)

120

In [18]:
# Calculating logarithms (the logarithm of 8 in base 2).
# The result is 3 because 2 to the power of 3 is 8.
math.log(8, 2)

3.0

In [19]:
# Calculating square roots.
math.sqrt(5)

2.23606797749979

In [20]:
# Calculating cosines.
math.cos(math.pi)

-1.0

# 4. Arrays

Up to now, we haven't done much that you couldn't do yourself by hand, without going through the trouble of learning Python.  Computers are most useful when you can use a small amount of code to *do the same action* to *many different things*.

For example, in the time it takes you to calculate the 18% tip on a restaurant bill, a laptop can calculate 18% tips for every restaurant bill paid by every human on Earth that day.  (That's if you're pretty fast at doing arithmetic in your head!)

**Arrays** are how we put many values in one place so that we can operate on them as a group. For example, if `billions_of_numbers` is an array of numbers, the expression

    .18 * billions_of_numbers

gives a new array of numbers that's the result of multiplying each number in `billions_of_numbers` by .18 (18%).  Arrays are not limited to numbers; we can also put all the words in a book into an array of strings.

Concretely, an array is a **collection of values of the same type**, like a column in an Excel spreadsheet. 

<img src="excel_array.jpg">

## 4.1. Making arrays
You can type in the data that goes in an array yourself, but normally we create arrays by loading them from an external source, like a data file.

First, execute the following cell so that all the names from the `datascience` module are available.

In [2]:
from datascience import *

Now, to create an array, call the function `make_array`.  Each argument you pass to `make_array` will be in the array it returns.  Run this cell to see an example:

In [3]:
make_array(0.125, 4.75, -1.3)

array([ 0.125,  4.75 , -1.3  ])

Each value in an array (in the above case, the numbers 0.125, 4.75, and -1.3) is called an *element* of that array.

Arrays themselves are also values, just like numbers and strings.  That means you can assign them names or use them as arguments to functions.

**Question 4.1.1.** Make an array containing the numbers 1, 2, and 3, in that order.  Name it `small_numbers`.

In [4]:
small_numbers = make_array(1, 2, 3)
small_numbers

array([1, 2, 3], dtype=int64)

**Question 4.1.2.** Make an array containing the numbers 0, 1, -1, $\pi$, and $e$, in that order.  Name it `interesting_numbers`.  *Hint:* How did you get the values $\pi$ and $e$ earlier?  You can refer to them in exactly the same way here.

In [6]:
import math
interesting_numbers = make_array(0, 1, -1, math.pi, math.e)
interesting_numbers

array([ 0.        ,  1.        , -1.        ,  3.14159265,  2.71828183])

**Question 4.1.3.** Make an array containing the five strings `"Hello"`, `","`, `" "`, `"world"`, and `"!"`.  (The third one is a single space inside quotes.)  Name it `hello_world_components`.

*Note:* If you print `hello_world_components`, you'll notice some extra information in addition to its contents: `dtype='<U5'`.  That's how NumPy indicates that the things in the array are strings.

In [7]:
hello_world_components = make_array("Hello", ",", " ", "world", "!")
hello_world_components

array(['Hello', ',', ' ', 'world', '!'], dtype='<U5')

The `join` method of a string takes an array of strings as its argument and puts all of the elements together into one string. Try it:

In [8]:
'°'.join(make_array('(╯', '□','）╯︵ ┻━┻'))

'(╯°□°）╯︵ ┻━┻'

**Question 4.1.4.** Assign `separator` to a string so that the name `hello` is bound to the string `'Hello, world!'` in the cell below.

In [9]:
separator = "hello"
hello = separator.join(hello_world_components)
hello

'Hellohello,hello helloworldhello!'

### 4.1.1.  `np.arange`
Arrays are provided by a package called [NumPy](http://www.numpy.org/) (pronounced "NUM-pie").  It is standard to rename `numpy` as `np` for brevity:

    import numpy as np

Very often in data science, we want to work with many numbers that are evenly spaced within some range.  NumPy provides a special function for this called `arange`.  `np.arange(start, stop, space)` produces an array with all the numbers starting at `start` and counting up by `space`, stopping before `stop` is reached.

For example, the value of `np.arange(1, 6, 2)` is an array with elements 1, 3, and 5 -- it starts at 1 and counts up by 2, then stops before 6.  In other words, it's equivalent to `make_array(1, 3, 5)`.

`np.arange(4, 9, 1)` is an array with elements 4, 5, 6, 7, and 8.  (It doesn't contain 9 because `np.arange` stops *before* the stop value is reached.)

**Question 4.1.1.1.** Import `numpy` as `np` and then use `np.arange` to create an array with the multiples of 99 from 0 up to (**and including**) 9999.  (So its elements are 0, 99, 198, 297, etc.)

In [10]:
import numpy as np
multiples_of_99 = np.arange(0, 1000, 99)
multiples_of_99

array([  0,  99, 198, 297, 396, 495, 594, 693, 792, 891, 990])

##### Temperature readings
NOAA (the US National Oceanic and Atmospheric Administration) operates weather stations that measure surface temperatures at different sites around the United States.  The hourly readings are [publicly available](http://www.ncdc.noaa.gov/qclcd/QCLCD?prior=N).

Suppose we download all the hourly data from the Blount County TN site for the month of December 2021.  To analyze the data, we want to know when each reading was taken, but we find that the data do not include the timestamps of the readings (the time at which each one was taken).

However, we know the first reading was taken at the first instant of December 2021 (midnight on December 1st) and each subsequent reading was taken exactly 1 hour after the last.

**Question 4.1.1.2.** Create an array of the *time, in seconds, since the start of the month* at which each hourly reading was taken.  Name it `collection_times`.

*Hint 1:* There were 31 days in December, which is equivalent to ($31 \times 24$) hours or ($31 \times 24 \times 60 \times 60$) seconds.  So your array should have $31 \times 24$ elements in it.

*Hint 2:* The `len` function works on arrays, too.  Check its length and make sure it has $31 \times 24$ elements.

In [12]:
import numpy as np
collection_times = np.arange(0, 31*24*60*60, 360)
collection_times

array([      0,     360,     720, ..., 2677320, 2677680, 2678040])

## 4.2. Working with single elements of arrays ("indexing")
Let's work with a more interesting dataset.  The next cell creates an array called `population` that includes estimated world populations in every year from **1950** to roughly the present.  (The estimates come from the [United Nations](https://population.un.org/wpp/Download/Standard/Population/).)

Rather than type in the data manually, you should have downloaded a file called `world_population.csv`. 

In [22]:
# Don't worry too much about what goes on in this cell.
from datascience import *
population = Table.read_table("world_population.csv").column("Population")
population

array([2536431018, 2584034227, 2630861690, 2677609061, 2724846754,
       2773019915, 2822443254, 2873306058, 2925686680, 2979576147,
       3034949715, 3091843513, 3150420761, 3211000946, 3273978272,
       3339583510, 3407922631, 3478770104, 3551599436, 3625680965,
       3700437042, 3775760030, 3851650588, 3927780519, 4003794178,
       4079480474, 4154666827, 4229505919, 4304533599, 4380506185,
       4458003466, 4536996619, 4617386526, 4699569187, 4784011517,
       4870921666, 4960568000, 5052521998, 5145425994, 5237441434,
       5327231041, 5414289383, 5498919893, 5581597598, 5663150428,
       5744212930, 5824891931, 5905045647, 5984794075, 6064239033,
       6143493806, 6222626531, 6301773172, 6381185141, 6461159391,
       6541906956, 6623517917, 6705946643, 6789088672, 6872766988,
       6956823588, 7041194168, 7125827957, 7210582041, 7295290759,
       7379796967, 7464021934, 7547858900, 7631091113, 7713468205,
       7794798729], dtype=int64)

Here's how we get the first element of `population`, which is the world population in the first year in the dataset, 1950.

In [23]:
population.item(0)

2536431018

The value of that expression is the number 2536431018 (around 2.5 billion), because that's the first thing in the array `population`.

Notice that we wrote `.item(0)`, not `.item(1)`, to get the first element.  This is by convention; 0 is called the *index* of the first item.  It's the number of elements that appear *before* that item.  So 3 is the index of the 4th item.

Here are some more examples.  In the examples, we've given names to the things we get out of `population`.  Read and run each cell.

In [24]:
# The third element in the array is the population
# in 1952.
population_1952 = population.item(2)
population_1952

2630861690

In [25]:
# The thirteenth element in the array is the population
# in 1962 (which is 1950 + 12).
population_1962 = population.item(12)
population_1962

3150420761

In [26]:
# The 71st element is the population in 2020.
population_2020 = population.item(70)
population_2020

7794798729

In [29]:
# The array has only 71 elements, so this doesn't work.
# (There's no element with 71 other elements before it.)
population_2021 = population.item(71)
population_2021

IndexError: index 71 is out of bounds for axis 0 with size 71

In [28]:
# Since make_array returns an array, we can call .item(3)
# on its output to get its 4th element, just like we
# "chained" together calls to the method "replace" earlier.
make_array(-1, -3, 4, -2).item(3)

-2

**Question 4.2.1.** Set `population_1973` to the world population in 1973, by getting the appropriate element from `population` using `item`.

In [30]:
population_1973 = population.item(23)
population_1973

3927780519

## 4.3. Doing something to every element of an array
Arrays are primarily useful for doing the same operation many times, so we don't often have to use `.item` and work with single elements.

##### Logarithms
Here is one simple question we might ask about world population:

> How big was the population in *orders of magnitude* in each year?

The logarithm function is one way of measuring how big a number is. The logarithm (base 10) of a number increases by 1 every time we multiply the number by 10. It's like a measure of how many decimal digits the number has, or how big it is in orders of magnitude.

We could try to answer our question like this, using the `log10` function from the `math` module and the `item` method you just saw:

In [31]:
population_1950_magnitude = math.log10(population.item(0))
population_1951_magnitude = math.log10(population.item(1))
population_1952_magnitude = math.log10(population.item(2))
population_1953_magnitude = math.log10(population.item(3))
...

Ellipsis

But this is tedious and doesn't really take advantage of the fact that we are using a computer.

Instead, NumPy provides its own version of `log10` that takes the logarithm of each element of an array.  It takes a single array of numbers as its argument.  It returns an array of the same length, where the first element of the result is the logarithm of the first element of the argument, and so on.

**Question 4.3.1.** Use it to compute the logarithms of the world population in every year.  Give the result (an array of 71 numbers) the name `population_magnitudes`.  Your code should be very short.

In [36]:
import numpy as np
population_magnitudes = np.log10(population)
population_magnitudes

array([9.40422306, 9.41229826, 9.42009802, 9.42774717, 9.43534208,
       9.44295299, 9.45062522, 9.45838189, 9.46622781, 9.47415449,
       9.4821515 , 9.4902175 , 9.49836856, 9.50664043, 9.51507579,
       9.52369231, 9.53248973, 9.54142573, 9.55042398, 9.55938959,
       9.56825302, 9.57700438, 9.58564688, 9.59414721, 9.60247174,
       9.61060486, 9.6185362 , 9.62628964, 9.6339261 , 9.6415243 ,
       9.6491404 , 9.65676846, 9.66439623, 9.67205805, 9.67979222,
       9.68761115, 9.69553141, 9.70350821, 9.71142134, 9.71911918,
       9.72650153, 9.73354146, 9.74027739, 9.74675852, 9.7530581 ,
       9.75923053, 9.76528787, 9.77122326, 9.77704921, 9.78277631,
       9.78841542, 9.79397374, 9.79946277, 9.80490135, 9.81031045,
       9.81570436, 9.82108872, 9.82646009, 9.83181148, 9.83713162,
       9.84241099, 9.84764632, 9.85283533, 9.85797032, 9.86304261,
       9.86804441, 9.87297291, 9.87782377, 9.88258664, 9.88724969,
       9.89180491])

<img src="array_logarithm.jpg">

This is called *elementwise* application of the function, since it operates separately on each element of the array it's called on.  The textbook's section on arrays has a useful list of NumPy functions that are designed to work elementwise, like `np.log10`.

##### Arithmetic
Arithmetic also works elementwise on arrays.  For example, you can divide all the population numbers by 1 billion to get numbers in billions:

In [35]:
population_in_billions = population / 1000000000
population_in_billions

array([2.53643102, 2.58403423, 2.63086169, 2.67760906, 2.72484675,
       2.77301991, 2.82244325, 2.87330606, 2.92568668, 2.97957615,
       3.03494972, 3.09184351, 3.15042076, 3.21100095, 3.27397827,
       3.33958351, 3.40792263, 3.4787701 , 3.55159944, 3.62568096,
       3.70043704, 3.77576003, 3.85165059, 3.92778052, 4.00379418,
       4.07948047, 4.15466683, 4.22950592, 4.3045336 , 4.38050618,
       4.45800347, 4.53699662, 4.61738653, 4.69956919, 4.78401152,
       4.87092167, 4.960568  , 5.052522  , 5.14542599, 5.23744143,
       5.32723104, 5.41428938, 5.49891989, 5.5815976 , 5.66315043,
       5.74421293, 5.82489193, 5.90504565, 5.98479407, 6.06423903,
       6.14349381, 6.22262653, 6.30177317, 6.38118514, 6.46115939,
       6.54190696, 6.62351792, 6.70594664, 6.78908867, 6.87276699,
       6.95682359, 7.04119417, 7.12582796, 7.21058204, 7.29529076,
       7.37979697, 7.46402193, 7.5478589 , 7.63109111, 7.7134682 ,
       7.79479873])

You can do the same with addition, subtraction, multiplication, and exponentiation (`**`). For example, you can calculate a tip on several restaurant bills at once (in this case just 3):

In [37]:
restaurant_bills = make_array(20.12, 39.90, 31.01)
print("Restaurant bills:\t", restaurant_bills)
tips = .2 * restaurant_bills
print("Tips:\t\t\t", tips)

Restaurant bills:	 [20.12 39.9  31.01]
Tips:			 [4.024 7.98  6.202]


<img src="array_multiplication.jpg">

**Question 4.3.2.** Suppose the total charge at a restaurant is the original bill plus the tip.  That means we can multiply the original bill by 1.2 to get the total charge.  Compute the total charge for each bill in `restaurant_bills`.

In [38]:
total_charges = make_array(1.2*restaurant_bills)
total_charges

array([[24.144, 47.88 , 37.212]])

**Question 4.3.3.** `more_restaurant_bills.csv` contains 100,000 bills!  Compute the total charge for each one.  How is your code different?

In [39]:
more_restaurant_bills = Table.read_table("more_restaurant_bills.csv").column("Bill")
more_total_charges = make_array(1.2*more_restaurant_bills)
more_total_charges

array([[20.244, 20.892, 12.216, ..., 19.308, 18.336, 35.664]])

The function `sum` takes a single array of numbers as its argument.  It returns the sum of all the numbers in that array (so it returns a single number, not an array).

**Question 4.3.4.** What was the sum of all the bills in `more_restaurant_bills`, *including tips*?

In [41]:
sum_of_bills = sum(1.2*more_restaurant_bills)
sum_of_bills

1795730.0640000193

**Question 4.3.5.** The powers of 2 ($2^0 = 1$, $2^1 = 2$, $2^2 = 4$, etc) arise frequently in computer science.  (For example, you may have noticed that storage on smartphones or USBs come in powers of 2, like 16 GB, 32 GB, or 64 GB.)  Use `np.arange` and the exponentiation operator `**` to compute the first 30 powers of 2, starting from `2^0`.

In [43]:
import numpy as np
powers_of_2 = 2**np.arange(0, 30, 1)
powers_of_2

array([        1,         2,         4,         8,        16,        32,
              64,       128,       256,       512,      1024,      2048,
            4096,      8192,     16384,     32768,     65536,    131072,
          262144,    524288,   1048576,   2097152,   4194304,   8388608,
        16777216,  33554432,  67108864, 134217728, 268435456, 536870912],
      dtype=int32)

## 

## 5. Submission

Congratulations, you're done with lab 3a! To submit this lab, please download your notebook as a .ipynb file and submit in Canvas under Lab 3a (under Assignments). To export, go to the toolbar at the top of this page, click File > Download as... > Notebook (.ipynb). Then, go to our class's Canvas page and upload your file under "Lab 3a". 

For easy identification, **please add your surname to the file**, as in: lab_03a_DS201_Name.ipynb